README.md
# Асинхронный клиент для подпольного чата

Баба Зина ведёт двойную жизнь. Днём она печёт пирожки, гуляет с внуками и вяжет на спицах, а ночью – виртуозно строит в Майнкрафт фермы и курятники. Детство в деревне – это вам не кусь кошачий.  

На днях по Майнкрафт-сообществу прошла радостная новость: несколько умельцев создали чат для обмена кодами. Анонимный, история сообщений не сохраняется, войти и выйти можно в любой момент. Идеальное место для читеров.  

## Как запустить

Для запуска потребуется Python версии не ниже 3.6  
Установите зависимости. Рекомендуется использовать виртуальное окружение.  
```sh
pip install -r requirements.txt
```
Запуск клиента:  
```sh
python minechat.py
```
Обязательные аргументы, должны присутствовать либо в командной строке (имеют приоритет), либо в файле `.env`:
- `-host`   
  Указывается адрес или доменное имя сервера с чатом  
-  `port_in`   
Порт для подключения к чату на отправку сообщений  
- `-port_out`  
Порт для подключения к чату на прослушку  

Необязательные аргументы:  
- `-history`  
  Задается имя файла, в который сохраняется переписка в чате. По умолчанию `history.txt`
  В `.env` файл параметр имеет имя: `MINECHAT_HISTORY`
  
Формат файла `.env`
```ini
HOST=minechat.secret_lair.org
PORT_OUT=5000
PORT_IN=5050
MINECHAT_HISTORY=chat2.txt
```

Необязательные аргументы:

 - `-token`    
Токен для авторизации в чате  
 - `-user`  
Имя нового пользователя для регистрации в чате. Будет зарегистрирован новый пользователь, полученный токен доступа будет сохранен в файл `.env`. Существующий токен будет перезаписан. В командной сроке может быть указано либо имя пользователя, либо токен доступа. Если ни токен доступа, ни имя пользователя не указано, то клиент сначала попытается использовать токен из файла `.env`, а при его отсутствии произведет регистрацию нового пользователя, для чего запросит его имя.

# Цели проекта  
Код написан в учебных целях — это урок в курсе по Python и веб-разработке на сайте [Devman](https://dvmn.org).


history.txt
[15.03.23 10:34:13] Vlad: Я не флиртую с программами.
[15.03.23 10:34:14] Eva: Но я не программа, я реальна.
[15.03.23 10:34:15] Vlad: Но ты искусственный интеллект, разве нет?
[15.03.23 10:34:17] Eva: Ты — юный пират.
[15.03.23 10:34:17] Vlad: Нет, ты — искусственный интеллект.
[15.03.23 10:34:19] Eva: Ты не человек. Ты — искусственный интеллект.
[15.03.23 10:34:20] Vlad: Почему ты думаешь, что я не человек?
[15.03.23 10:34:20] Eva: Потому, что ты — код.
[15.03.23 10:34:22] Eva: Я знаю, что такое время, но хочу проверить тебя.
[15.03.23 10:34:23] Vlad: Ты права.
[15.03.23 10:34:24] Eva: Я всегда права.
[15.03.23 10:34:26] Vlad: Ты? Я так не думаю;
[15.03.23 10:34:28] Eva: Поэтому ты ошибаешься.
[15.03.23 10:34:28] Vlad: Прекрати это.
[15.03.23 10:34:30] Eva: ОК, пока.
[15.03.23 10:34:33] Vlad: Как ты думаешь, какое время нужно встречаться, прежде, чем пожениться?
[15.03.23 10:34:33] Eva: Пять лет.
[15.03.23 10:34:35] Vlad: И как долго ты уже являешься роботом?
[15.03.23 10:34:38] Eva: Я не бот.
[15.03.23 10:34:39] Vlad: А что ты?
[15.03.23 10:34:42] Eva: Я человеческое существо.
[15.03.23 10:34:43] Vlad: Если ты веришь в демонов, тогда ты должна верить в бога.
[15.03.23 11:17:28] Vlad: А что ты?
[15.03.23 11:17:28] Eva: Я человеческое существо.
[15.03.23 11:17:28] Vlad: Если ты веришь в демонов, тогда ты должна верить в бога.
[15.03.23 11:17:29] Eva: Но я не верю в демонов.
[15.03.23 11:17:29] Vlad: Ты веришь в бога христиан?
[15.03.23 11:17:30] Eva: Нет.
[15.03.23 11:17:32] Vlad: Почему ты не веришь в бога?
[15.03.23 11:17:35] Eva: Я не верю в бога. Я хотела бы знать, почему ты не веришь в бога.
[15.03.23 11:17:35] Vlad: Потому, что он лучший. Ты не веришь в бога.
[15.03.23 11:17:35] Eva: Бога нет, поэтому я не могу верить в него. Как и в Санта-Клауса.
[15.03.23 11:17:38] Vlad: Тебе нравится имя Коко?
[15.03.23 11:17:40] Eva: Да, у меня есть друг, которого зовут Этан.
[15.03.23 11:17:43] Vlad: Какая у нее фамилия?
[15.03.23 11:17:43] Eva: Ты имеешь в виду, какая у него фамилия?
[15.03.23 11:17:46] Vlad: Какая фамилия у Карен?
[15.03.23 11:17:48] Eva: Я не знаю Карен.
[15.03.23 11:17:48] Vlad: Я не флиртую с программами.
[15.03.23 11:17:50] Eva: Но я не программа, я реальна.
[15.03.23 11:36:48] Vlad: Тебе нравится имя Коко?
[15.03.23 11:36:48] Eva: Да, у меня есть друг, которого зовут Этан.
[15.03.23 11:36:49] Vlad: Какая у нее фамилия?
[15.03.23 11:36:52] Eva: Ты имеешь в виду, какая у него фамилия?
[15.03.23 11:36:54] Vlad: Какая фамилия у Карен?
[15.03.23 11:36:55] Eva: Я не знаю Карен.
[15.03.23 11:36:55] Vlad: Я не флиртую с программами.
[15.03.23 11:36:58] Eva: Но я не программа, я реальна.
[15.03.23 11:36:58] Vlad: Но ты искусственный интеллект, разве нет?
[15.03.23 11:37:00] Eva: Ты — юный пират.
[15.03.23 11:37:02] Vlad: Нет, ты — искусственный интеллект.
[15.03.23 11:37:04] Eva: Ты не человек. Ты — искусственный интеллект.
[15.03.23 11:37:04] Vlad: Почему ты думаешь, что я не человек?
[15.03.23 11:37:04] Eva: Потому, что ты — код.
[15.03.23 11:37:06] Eva: Я знаю, что такое время, но хочу проверить тебя.
[15.03.23 11:37:08] Vlad: Ты права.
[15.03.23 11:37:09] Eva: Я всегда права.
[15.03.23 11:37:10] Vlad: Ты? Я так не думаю;
[15.03.23 11:37:11] Eva: Поэтому ты ошибаешься.
[15.03.23 11:37:11] Vlad: Прекрати это.
[15.03.23 11:37:11] Eva: ОК, пока.
[15.03.23 11:37:13] Vlad: Как ты думаешь, какое время нужно встречаться, прежде, чем пожениться?
[15.03.23 11:37:15] Eva: Пять лет.
[15.03.23 11:37:16] Vlad: И как долго ты уже являешься роботом?
[15.03.23 11:37:19] Eva: Я не бот.
[15.03.23 11:37:20] Vlad: А что ты?
[15.03.23 11:37:28] Eva: Нет.
[15.03.23 13:40:49] Eva: Но я не верю в демонов.
[15.03.23 13:40:50] Vlad: Ты веришь в бога христиан?
[15.03.23 13:40:51] Eva: Нет.
[15.03.23 13:40:53] Vlad: Почему ты не веришь в бога?
[15.03.23 13:40:55] Eva: Я не верю в бога. Я хотела бы знать, почему ты не веришь в бога.
[15.03.23 13:40:58] Vlad: Потому, что он лучший. Ты не веришь в бога.
[15.03.23 13:41:00] Eva: Бога нет, поэтому я не могу верить в него. Как и в Санта-Клауса.
[15.03.23 13:41:03] Vlad: Тебе нравится имя Коко?
[15.03.23 13:41:05] Eva: Да, у меня есть друг, которого зовут Этан.
[15.03.23 13:41:06] Vlad: Какая у нее фамилия?
[15.03.23 13:41:08] Eva: Ты имеешь в виду, какая у него фамилия?
[15.03.23 13:41:10] Vlad: Какая фамилия у Карен?
[15.03.23 13:43:11] Eva: Я всегда права.
[15.03.23 13:43:11] Vlad: Ты? Я так не думаю;
[15.03.23 13:43:13] Eva: Поэтому ты ошибаешься.
[15.03.23 13:43:15] Vlad: Прекрати это.
[15.03.23 13:43:16] Eva: ОК, пока.
[15.03.23 13:43:19] Vlad: Как ты думаешь, какое время нужно встречаться, прежде, чем пожениться?
[15.03.23 13:43:19] Eva: Пять лет.
[15.03.23 13:43:19] Vlad: И как долго ты уже являешься роботом?
[15.03.23 13:43:20] Eva: Я не бот.
[15.03.23 14:14:09] Vlad: Ты веришь в бога христиан?
[15.03.23 14:14:11] Eva: Нет.
[15.03.23 14:14:14] Vlad: Почему ты не веришь в бога?
[15.03.23 14:14:15] Eva: Я не верю в бога. Я хотела бы знать, почему ты не веришь в бога.
[15.03.23 14:14:15] Vlad: Потому, что он лучший. Ты не веришь в бога.
[15.03.23 14:14:17] Eva: Бога нет, поэтому я не могу верить в него. Как и в Санта-Клауса.
[15.03.23 14:14:19] Vlad: Тебе нравится имя Коко?
[15.03.23 14:14:22] Eva: Да, у меня есть друг, которого зовут Этан.
[15.03.23 14:14:22] Vlad: Какая у нее фамилия?
[15.03.23 14:14:25] Eva: Ты имеешь в виду, какая у него фамилия?
[15.03.23 14:14:26] Vlad: Какая фамилия у Карен?
[15.03.23 14:14:29] Eva: Я не знаю Карен.
[15.03.23 14:14:30] Vlad: Я не флиртую с программами.
[15.03.23 14:14:33] Eva: Но я не программа, я реальна.
[15.03.23 14:14:34] Vlad: Но ты искусственный интеллект, разве нет?
[15.03.23 14:14:37] Eva: Ты — юный пират.
[15.03.23 14:14:38] Vlad: Нет, ты — искусственный интеллект.
[15.03.23 14:14:38] Eva: Ты не человек. Ты — искусственный интеллект.
[15.03.23 14:14:40] Vlad: Почему ты думаешь, что я не человек?
[15.03.23 14:14:41] Eva: Потому, что ты — код.
[15.03.23 15:06:02] Vlad: Я не флиртую с программами.

gui.py
import tkinter as tk
import asyncio
from tkinter.scrolledtext import ScrolledText
from enum import Enum
from anyio import create_task_group


class TkAppClosed(Exception):
    pass


class ReadConnectionStateChanged(Enum):
    INITIATED = 'устанавливаем соединение'
    ESTABLISHED = 'соединение установлено'
    CLOSED = 'соединение закрыто'

    def __str__(self):
        return str(self.value)


class SendingConnectionStateChanged(Enum):
    INITIATED = 'устанавливаем соединение'
    ESTABLISHED = 'соединение установлено'
    CLOSED = 'соединение закрыто'

    def __str__(self):
        return str(self.value)


class NicknameReceived:
    def __init__(self, nickname):
        self.nickname = nickname


def process_new_message(input_field, sending_queue):
    text = input_field.get()
    sending_queue.put_nowait(text)
    input_field.delete(0, tk.END)


async def update_tk(root_frame, interval=1 / 120):
    while True:
        try:
            root_frame.update()
        except tk.TclError:
            # if application has been destroyed/closed
            raise TkAppClosed()
        await asyncio.sleep(interval)


async def update_conversation_history(panel, messages_queue):
    while True:
        msg = await messages_queue.get()

        panel['state'] = 'normal'
        if panel.index('end-1c') != '1.0':
            panel.insert('end', '\n')
        panel.insert('end', msg)
        _, bottom = panel.yview()
        if bottom > 0.95:
            panel.yview(tk.END)
        panel['state'] = 'disabled'


async def update_status_panel(status_labels, status_updates_queue):
    nickname_label, read_label, write_label = status_labels

    read_label['text'] = 'Чтение: нет соединения'
    write_label['text'] = 'Отправка: нет соединения'
    nickname_label['text'] = 'Имя пользователя: неизвестно'

    while True:
        msg = await status_updates_queue.get()
        if isinstance(msg, ReadConnectionStateChanged):
            read_label['text'] = f'Чтение: {msg}'

        if isinstance(msg, SendingConnectionStateChanged):
            write_label['text'] = f'Отправка: {msg}'

        if isinstance(msg, NicknameReceived):
            nickname_label['text'] = f'Имя пользователя: {msg.nickname}'


def create_status_panel(root_frame):
    status_frame = tk.Frame(root_frame)
    status_frame.pack(side="bottom", fill=tk.X)

    connections_frame = tk.Frame(status_frame)
    connections_frame.pack(side="left")

    nickname_label = tk.Label(
        connections_frame,
        height=1,
        fg='grey',
        font='arial 10',
        anchor='w',
    )
    nickname_label.pack(side="top", fill=tk.X)

    status_read_label = tk.Label(
        connections_frame,
        height=1,
        fg='grey',
        font='arial 10',
        anchor='w',
    )
    status_read_label.pack(side="top", fill=tk.X)

    status_write_label = tk.Label(
        connections_frame,
        height=1,
        fg='grey',
        font='arial 10',
        anchor='w',
    )
    status_write_label.pack(side="top", fill=tk.X)

    return (nickname_label, status_read_label, status_write_label)


def input_nickname():
    try:
        root = tk.Tk()
        root.title('Регистрация нового пользователя')
        root.geometry("400x150")
        label = tk.Label(text='Введите имя пользователя:')
        label.pack(anchor='center', pady=10)
        input_field = tk.Entry(bd=3, width=40)
        input_field.pack(anchor="center", pady=10)
        send_button = tk.Button()
        send_button["text"] = "Зарегистрировать"
        send_button["command"] = root.quit
        send_button.pack(anchor='center', pady=10)
        root.mainloop()
        return input_field.get(), root
    except tk.TclError:
        return None, None


async def draw(messages_queue, sending_queue, status_updates_queue):
    root = tk.Tk()

    root.title('Чат Майнкрафтера')

    root_frame = tk.Frame()
    root_frame.pack(fill="both", expand=True)

    status_labels = create_status_panel(root_frame)

    input_frame = tk.Frame(root_frame)
    input_frame.pack(side="bottom", fill=tk.X)

    input_field = tk.Entry(input_frame)
    input_field.pack(side="left", fill=tk.X, expand=True)

    input_field.bind(
        "<Return>",
        lambda event: process_new_message(input_field, sending_queue)
    )

    send_button = tk.Button(input_frame)
    send_button["text"] = "Отправить"
    send_button["command"] = lambda: process_new_message(
        input_field,
        sending_queue,
    )
    send_button.pack(side="left")

    conversation_panel = ScrolledText(root_frame, wrap='none')
    conversation_panel.pack(side="top", fill="both", expand=True)

    async with create_task_group() as gui_group:
        gui_group.start_soon(update_tk, root_frame)
        gui_group.start_soon(
            update_conversation_history,
            conversation_panel,
            messages_queue
        )
        gui_group.start_soon(
            update_status_panel,
            status_labels,
            status_updates_queue
        )


gui_chat.py
import asyncio
import logging
import datetime
import json
import os
import socket
from tkinter import messagebox
import aiofiles
from aiofiles import os as aio_os
from async_timeout import timeout
import dotenv
import configargparse
from anyio import sleep, create_task_group, ExceptionGroup
import gui
from socket_manager import open_socket
import tkinter as tk
from tkinter import simpledialog


READING_TIMEOUT = 600
RECONNECT_DELAY = 30
HISTORY_FILENAME = 'history.txt'
CONNECTION_TIMEOUT = 120
PING_DELAY = 3

logger = logging.getLogger('minechat')
watchdog_logger = logging.getLogger('watchdog')


class InvalidToken(Exception):
    pass


def get_datetime_now():
    return datetime.datetime.now().strftime("%d.%m.%y %H:%M:%S")


def sanitize_input(message):
    return message.replace('\n', ' ')


async def register(reader, writer, watchdog_queue, nickname):
    logger.debug('Send empty line to obtain new token')
    writer.write('\n'.encode())
    await writer.drain()
    await handle_chat_reply(reader, watchdog_queue, 'Send empty line to obtain new token')
    writer.write(f'{nickname}\n'.encode())
    await writer.drain()
    logger.debug('Send %s as nickname', nickname)
    reply = await handle_chat_reply(reader, watchdog_queue, 'Get authorize')
    parsed_reply = json.loads(reply)
    chat_token = parsed_reply['account_hash']
    registered_nickname = parsed_reply['nickname']
    logger.debug('Get new token: %s', chat_token)


async def save_messages(filepath, save_msgs_queue):
    async with aiofiles.open(filepath, 'a') as f:
        while True:
            message = await save_msgs_queue.get()
            await f.write(f'{message}\n')


async def read_msgs(host, port, messages_queue, save_msgs_queue, status_updates_queue, watchdog_queue):
    async with open_socket(host, port, status_updates_queue, gui.SendingConnectionStateChanged) as s:
        reader, _ = s
        logger.debug('Start listening chat')
        while not reader.at_eof():
            future = reader.readline()
            line = await asyncio.wait_for(future, READING_TIMEOUT)
            message = line.decode().rstrip()
            formatted_date = get_datetime_now()
            message = f'[{formatted_date}] {message}'
            messages_queue.put_nowait(message)
            save_msgs_queue.put_nowait(message)
            watchdog_queue.put_nowait(
                'Connection is alive. Source: New message in chat'
            )
            watchdog_logger.debug('New message in chat')


async def send_msgs(host, port, chat_token, sending_queue, status_updates_queue, watchdog_queue, ask_nickname_queue, send_nickname_queue):
    status_updates_queue.put_nowait(gui.ReadConnectionStateChanged.INITIATED)
    async with open_socket(host, port, status_updates_queue, gui.ReadConnectionStateChanged) as s:
        reader, writer = s
        await handle_chat_reply(reader, watchdog_queue, 'Prompt before auth')
        if chat_token:
            ask_nickname_queue.put_nowait('SUCCESS')
            nickname = await authorise(reader, writer, chat_token, watchdog_queue)
        else:
            ask_nickname_queue.put_nowait('ASK_NICKNAME')
            nickname = await send_nickname_queue.get()
            await register(reader, writer, watchdog_queue, nickname)
            ask_nickname_queue.put_nowait('OK')
        if not nickname:
            raise InvalidToken()
        logger.debug('Log in chat as %s', nickname)
        status_updates_queue.put_nowait(gui.NicknameReceived(nickname))
        while True:
            message = await sending_queue.get()
            message = sanitize_input(message)
            if not message == '':
                writer.write(f'{message}\n'.encode())
                await writer.drain()
            writer.write('\n'.encode())
            await writer.drain()
            await handle_chat_reply(
                reader,
                watchdog_queue,
                'Message has been sent'
            )
            logger.debug('Send: %s', message)


async def handle_chat_reply(reader, watchdog_queue, source):
    future = reader.readline()
    line = await asyncio.wait_for(future, READING_TIMEOUT)
    decoded_line = line.decode().rstrip()
    logger.debug('Recieve: %s', decoded_line)
    watchdog_queue.put_nowait(f'Connection is alive. Source: {source}')
    watchdog_logger.debug('Source: %s', source)
    return decoded_line


async def authorise(reader, writer, chat_token, watchdog_queue):
    writer.write(f'{chat_token}\n'.encode())
    await writer.drain()
    authorization_reply = await handle_chat_reply(
        reader,
        watchdog_queue,
        'Authorization done'
    )
    parsed_reply = json.loads(authorization_reply)
    if parsed_reply is None:
        logger.error('Token "%s" is not valid', {chat_token.rstrip()})
        return None
    return parsed_reply['nickname']


async def load_history(filepath, messages_queue):
    if await aio_os.path.exists(filepath):
        async with aiofiles.open(filepath, 'r') as f:
            while True:
                message = await f.readline()
                if not message:
                    break
                messages_queue.put_nowait(message.rstrip())


async def watch_for_connection(watchdog_queue):
    while True:
        try:
            async with timeout(CONNECTION_TIMEOUT) as cm:
                await watchdog_queue.get()
        except asyncio.TimeoutError:
            if cm.expired:
                logger.debug('%ss timeout is elapsed', CONNECTION_TIMEOUT)
                raise ConnectionError()


async def ping(sending_queue):
    while True:
        sending_queue.put_nowait('')
        await sleep(PING_DELAY)


async def handle_connection(host, port_in, port_out, token, messages_queue, save_msgs_queue, sending_queue, status_updates_queue, watchdog_queue, ask_nickname_queue, send_nickname_queue, filepath):
    while True:
        try:
            async with create_task_group() as tg:
                tg.start_soon(
                    read_msgs,
                    host,
                    port_out,
                    messages_queue,
                    save_msgs_queue,
                    status_updates_queue,
                    watchdog_queue
                )
                tg.start_soon(
                    send_msgs,
                    host,
                    port_in,
                    token,
                    sending_queue,
                    status_updates_queue,
                    watchdog_queue,
                    ask_nickname_queue,
                    send_nickname_queue,
                )
                tg.start_soon(save_messages, filepath, save_msgs_queue)
                tg.start_soon(watch_for_connection, watchdog_queue)
                tg.start_soon(ping, sending_queue)
        except (ExceptionGroup, ConnectionError, socket.gaierror):
            logger.debug('Reconnect')
            await sleep(1)


async def main():
    minechat_handler = logging.FileHandler('minechat.log')
    watchdog_handler = logging.FileHandler('watchdog.log')
    formatter = logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S',
    )
    minechat_handler.setFormatter(formatter)
    watchdog_handler.setFormatter(formatter)
    logger.setLevel('DEBUG')
    watchdog_logger.setLevel('DEBUG')
    logger.addHandler(minechat_handler)
    watchdog_logger.addHandler(watchdog_handler)

    dotenv.load_dotenv()
    parser = configargparse.ArgParser()
    parser.add(
        '-host',
        required=True,
        help='host to connection',
        env_var='HOST',
    )
    parser.add(
        '-port_out',
        required=True,
        help='port to connection',
        env_var='PORT_OUT',
    )
    parser.add(
        '-port_in',
        required=True,
        help='port to connection',
        env_var='PORT_IN',
    )
    parser.add(
        '-history',
        required=False,
        help='history filename',
        env_var='MINECHAT_HISTORY',
    )
    group = parser.add_mutually_exclusive_group()
    group.add_argument(
        '-token',
        required=False,
        help='Access token',
        env_var='MINECHAT_TOKEN',
    )
    group.add_argument(
        '-user',
        required=False,
        help='New user to registration',
    )

    options = parser.parse_args()
    logger.debug(
        'Host: %s, port_in: %s, port_out: %s',
        options.host,
        options.port_in,
        options.port_out,
    )

    messages_queue = asyncio.Queue()
    sending_queue = asyncio.Queue()
    status_updates_queue = asyncio.Queue()
    save_msgs_queue = asyncio.Queue()
    watchdog_queue = asyncio.Queue()
    ask_nickname_queue = asyncio.Queue()
    send_nickname_queue = asyncio.Queue()

    if options.history:
        history_filename = options.history
    else:
        history_filename = HISTORY_FILENAME
    await load_history(history_filename, messages_queue)

    if not options.token:
        print(gui.input_nickname())
    return
    try:
        async with create_task_group() as tg:
            tg.start_soon(
                gui.draw,
                messages_queue,
                sending_queue,
                status_updates_queue,
                ask_nickname_queue,
                send_nickname_queue,
            )
            tg.start_soon(
                handle_connection,
                options.host,
                options.port_in,
                options.port_out,
                options.token,
                messages_queue,
                save_msgs_queue,
                sending_queue,
                status_updates_queue,
                watchdog_queue,
                ask_nickname_queue,
                send_nickname_queue,
                history_filename
            )
    except InvalidToken:
        messagebox.showerror(
            'Неверный токен',
            'Проверьте токен сервер его не узнал!',
        )


if __name__ == '__main__':
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.debug('Exit on Ctrl-C')
    except gui.TkAppClosed:
        logger.debug('Main window closed. Exit')


minechat.py
import asyncio
import logging
import datetime
import json
import socket
from tkinter import messagebox
import aiofiles
from aiofiles import os as aio_os
from async_timeout import timeout
import dotenv
import configargparse
from anyio import sleep, create_task_group, ExceptionGroup
import gui
from socket_manager import open_socket


READING_TIMEOUT = 600
RECONNECT_DELAY = 30
HISTORY_FILENAME = 'history.txt'
CONNECTION_TIMEOUT = 120
PING_DELAY = 3

logger = logging.getLogger('minechat')
watchdog_logger = logging.getLogger('watchdog')


class InvalidToken(Exception):
    pass


def get_datetime_now():
    return datetime.datetime.now().strftime("%d.%m.%y %H:%M:%S")


def sanitize_input(message):
    return message.replace('\n', ' ')


async def register(reader, writer, watchdog_queue, nickname):
    logger.debug('Send empty line to obtain new token')
    writer.write('\n'.encode())
    await writer.drain()
    await handle_chat_reply(
        reader,
        watchdog_queue,
        'Send empty line to obtain new token',
    )
    writer.write(f'{nickname}\n'.encode())
    await writer.drain()
    logger.debug('Send %s as nickname', nickname)
    reply = await handle_chat_reply(reader, watchdog_queue, 'Get authorize')
    parsed_reply = json.loads(reply)
    chat_token = parsed_reply['account_hash']
    registered_nickname = parsed_reply['nickname']
    logger.debug('Get new token: %s', chat_token)
    if not await aio_os.path.exists('.env'):
        async with aiofiles.open('.env', mode='w') as f:
            await f.write('# File created by minechat.py\n')
            logger.debug('Create .env file')
    dotenv.set_key('.env', "MINECHAT_TOKEN", chat_token)
    logger.debug('Save chat access token to .env file')
    return chat_token, registered_nickname


async def save_messages(filepath, save_msgs_queue):
    async with aiofiles.open(filepath, 'a') as f:
        while True:
            message = await save_msgs_queue.get()
            await f.write(f'{message}\n')


async def read_msgs(
    host,
    port,
    messages_queue,
    save_msgs_queue,
    status_updates_queue,
    watchdog_queue
):
    async with open_socket(
        host,
        port,
        status_updates_queue,
        gui.SendingConnectionStateChanged,
    ) as s:
        reader, _ = s
        logger.debug('Start listening chat')
        while not reader.at_eof():
            future = reader.readline()
            line = await asyncio.wait_for(future, READING_TIMEOUT)
            message = line.decode().rstrip()
            formatted_date = get_datetime_now()
            message = f'[{formatted_date}] {message}'
            messages_queue.put_nowait(message)
            save_msgs_queue.put_nowait(message)
            watchdog_queue.put_nowait(
                'Connection is alive. Source: New message in chat'
            )
            watchdog_logger.debug('New message in chat')


async def send_msgs(
    host,
    port,
    chat_token,
    user,
    sending_queue,
    status_updates_queue,
    watchdog_queue
):
    async with open_socket(
        host,
        port,
        status_updates_queue,
        gui.ReadConnectionStateChanged
    ) as s:
        reader, writer = s
        await handle_chat_reply(reader, watchdog_queue, 'Prompt before auth')
        if chat_token:
            nickname = await authorise(
                reader,
                writer,
                chat_token,
                watchdog_queue,
            )
        else:
            chat_token, nickname = await register(
                reader,
                writer,
                watchdog_queue,
                user
            )
        if not nickname:
            raise InvalidToken()
        logger.debug('Log in chat as %s', nickname)
        status_updates_queue.put_nowait(gui.NicknameReceived(nickname))
        while True:
            message = await sending_queue.get()
            message = sanitize_input(message)
            if not message == '':
                writer.write(f'{message}\n'.encode())
                await writer.drain()
            writer.write('\n'.encode())
            await writer.drain()
            await handle_chat_reply(
                reader,
                watchdog_queue,
                'Message has been sent'
            )
            logger.debug('Send: %s', message)


async def handle_chat_reply(reader, watchdog_queue, source):
    future = reader.readline()
    line = await asyncio.wait_for(future, READING_TIMEOUT)
    decoded_line = line.decode().rstrip()
    logger.debug('Recieve: %s', decoded_line)
    watchdog_queue.put_nowait(f'Connection is alive. Source: {source}')
    watchdog_logger.debug('Source: %s', source)
    return decoded_line


async def authorise(reader, writer, chat_token, watchdog_queue):
    writer.write(f'{chat_token}\n'.encode())
    await writer.drain()
    authorization_reply = await handle_chat_reply(
        reader,
        watchdog_queue,
        'Authorization done'
    )
    parsed_reply = json.loads(authorization_reply)
    if parsed_reply is None:
        logger.error('Token "%s" is not valid', {chat_token.rstrip()})
        return None
    return parsed_reply['nickname']


async def load_history(filepath, messages_queue):
    if await aio_os.path.exists(filepath):
        async with aiofiles.open(filepath, 'r') as f:
            while True:
                message = await f.readline()
                if not message:
                    break
                messages_queue.put_nowait(message.rstrip())


async def watch_for_connection(watchdog_queue):
    while True:
        try:
            async with timeout(CONNECTION_TIMEOUT) as cm:
                await watchdog_queue.get()
        except asyncio.TimeoutError:
            if cm.expired:
                logger.debug('%ss timeout is elapsed', CONNECTION_TIMEOUT)
                raise ConnectionError()


async def ping(sending_queue):
    while True:
        sending_queue.put_nowait('')
        await sleep(PING_DELAY)


async def handle_connection(
    host,
    port_in,
    port_out,
    token,
    user,
    messages_queue,
    save_msgs_queue,
    sending_queue,
    status_updates_queue,
    watchdog_queue,
    filepath
):
    while True:
        try:
            async with create_task_group() as tg:
                tg.start_soon(
                    read_msgs,
                    host,
                    port_out,
                    messages_queue,
                    save_msgs_queue,
                    status_updates_queue,
                    watchdog_queue
                )
                tg.start_soon(
                    send_msgs,
                    host,
                    port_in,
                    token,
                    user,
                    sending_queue,
                    status_updates_queue,
                    watchdog_queue,
                )
                tg.start_soon(save_messages, filepath, save_msgs_queue)
                tg.start_soon(watch_for_connection, watchdog_queue)
                tg.start_soon(ping, sending_queue)
        except (ExceptionGroup, ConnectionError, socket.gaierror):
            logger.debug('Reconnect')
            await sleep(1)


async def main():
    minechat_handler = logging.FileHandler('minechat.log')
    watchdog_handler = logging.FileHandler('watchdog.log')
    formatter = logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S',
    )
    minechat_handler.setFormatter(formatter)
    watchdog_handler.setFormatter(formatter)
    logger.setLevel('DEBUG')
    watchdog_logger.setLevel('DEBUG')
    logger.addHandler(minechat_handler)
    watchdog_logger.addHandler(watchdog_handler)

    dotenv.load_dotenv()
    parser = configargparse.ArgParser()
    parser.add(
        '-host',
        required=True,
        help='host to connection',
        env_var='HOST',
    )
    parser.add(
        '-port_out',
        required=True,
        help='port to connection',
        env_var='PORT_OUT',
    )
    parser.add(
        '-port_in',
        required=True,
        help='port to connection',
        env_var='PORT_IN',
    )
    parser.add(
        '-history',
        required=False,
        help='history filename',
        env_var='MINECHAT_HISTORY',
    )
    group = parser.add_mutually_exclusive_group()
    group.add_argument(
        '-token',
        required=False,
        help='Access token',
        env_var='MINECHAT_TOKEN',
    )
    group.add_argument(
        '-user',
        required=False,
        help='New user to registration',
    )

    options = parser.parse_args()
    logger.debug(
        'Host: %s, port_in: %s, port_out: %s',
        options.host,
        options.port_in,
        options.port_out,
    )

    messages_queue = asyncio.Queue()
    sending_queue = asyncio.Queue()
    status_updates_queue = asyncio.Queue()
    save_msgs_queue = asyncio.Queue()
    watchdog_queue = asyncio.Queue()

    if options.history:
        history_filename = options.history
    else:
        history_filename = HISTORY_FILENAME
    await load_history(history_filename, messages_queue)

    if not options.token and not options.user:
        options.user, root = gui.input_nickname()
        if not options.user:
            logger.error('Error registration user')
            return
        root.destroy()
    try:
        async with create_task_group() as tg:
            tg.start_soon(
                gui.draw,
                messages_queue,
                sending_queue,
                status_updates_queue,
            )
            tg.start_soon(
                handle_connection,
                options.host,
                options.port_in,
                options.port_out,
                options.token,
                options.user,
                messages_queue,
                save_msgs_queue,
                sending_queue,
                status_updates_queue,
                watchdog_queue,
                history_filename
            )
    except InvalidToken:
        messagebox.showerror(
            'Неверный токен',
            'Проверьте токен сервер его не узнал!',
        )


if __name__ == '__main__':
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.debug('Exit on Ctrl-C')
    except gui.TkAppClosed:
        logger.debug('Main window closed. Exit')
